web/modules/custom/joinup_core/src/EventSubscriber/GroupInvitationSubscriber.php

```diff
@@ -0,0 +1,142 @@
+    $this->membershipManager = $membershipManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents(): array {
+    $events[InvitationEvents::ACCEPT_INVITATION_EVENT] = ['acceptInvitation'];
+    $events[InvitationEvents::REJECT_INVITATION_EVENT] = ['rejectInvitation'];
+
+    return $events;
+  }
+
+  /**
+   * Accepts invitations to join a group.
+   *
+   * @param \Drupal\joinup_invite\Event\InvitationEventInterface $event
+   *   The event that was fired.
+   */
+  public function acceptInvitation(InvitationEventInterface $event): void {
+    $invitation = $event->getInvitation();
+
+    // Ignore invitations to other content entities.
+    if ($invitation->bundle() !== 'group') {
+      return;
+    }
+
+    $user = $invitation->getRecipient();
+    $group = $invitation->getEntity();
+
+    if ($this->getMembership($group, $user)) {
+      drupal_set_message($this->t('There is no action pending for this user.'));
+      // If a user already has a membership and is here, it means that there was
+      // a leftover pending invitation (the subscriber is called after a pending
+      // invitation is accepted). In that case, delete the invitation. The
+      // reason behind it is that if the invitation is not deleted, it might
+      // falsely seem that the membership exists due to an accepted invitation.
+      $invitation->delete();
+    }
+    else {
+      $membership = $this->membershipManager->createMembership($group, $user);
+      if (!empty($invitation->get('invitation_role')->first())) {
+        $role = $invitation->get('invitation_role')->first()->entity;
+        $membership->addRole($role);
+      }
+      // Disable notifications related to memberships.
+      $membership->skip_notification = TRUE;
+      $membership->setState(OgMembershipInterface::STATE_ACTIVE)->save();
+      $facilitator_id = $group->getEntityTypeId() . '-' . $group->bundle() . '-facilitator';
+      $role_argument = $membership->hasRole($facilitator_id) ? $this->t('facilitator') : $this->t('member');
+      drupal_set_message($this->t('You are now a @role of the "@title" @bundle.', [
+        '@role' => $role_argument,
+        '@title' => $group->label(),
+        '@bundle' => $group->get('rid')->entity->getSingularLabel(),
+      ]));
+    }
+  }
+
+  /**
+   * Rejects invitations to join a group.
+   *
+   * @param \Drupal\joinup_invite\Event\InvitationEventInterface $event
+   *   The event that was fired.
+   */
+  public function rejectInvitation(InvitationEventInterface $event): void {
+    $invitation = $event->getInvitation();
+
+    // Ignore invitations to other content entities.
+    if ($invitation->bundle() !== 'group') {
+      return;
+    }
+
+    $user = $invitation->getRecipient();
+    $group = $invitation->getEntity();
+
+    // If there is already a membership, it means that it has been created
+    // somewhere else in the meantime.
+    if (!empty($this->getMembership($group, $user))) {
```

Comment:
As for accept, I think we should make it clear. But I have no strong opinion on what we should do for each case. Maybe we can clarify this together?

Reply:
> As for accept
what do you mean with that? Since we refactored the code and there will be no membership created, this is a case where for some reason, a membership exists and the invitation was still pending.
This is a failsafe actually. It should not happen according to our code.

Reply:
Then for both, accept & reject, we should throw probably an exception?